ProjFS.Mac/PrjFSLib/PrjFSLib.cpp

```diff
@@ -758,37 +759,42 @@ static void HandleKernelRequest(void* messageMemory, uint32_t messageSize)
                 requestHeader,
                 relativePath,
                 absolutePath,
+                nullptr, /* relativeFromPath */
                 requestHeader->messageType == MessageType_KtoU_NotifyDirectoryPreDelete,  // isDirectory
                 KUMessageTypeToNotificationType(static_cast<MessageType>(requestHeader->messageType)));
             break;
         }
         
         case MessageType_KtoU_NotifyFileHardLinkCreated:
         {
-#if DEBUG
-            // TODO(Mac): Move the following line out of the DEBUG block once we actually need the information. Currently just causes warning-as-error in release build.
-            const char* fromPath = request.paths[MessagePath_From];
-            const char* relativeFromPath = fromPath != nullptr ? GetRelativePath(fromPath, s_virtualizationRootFullPath.c_str()) : nullptr;
-
-            cout << "PrjFSLib.HandleKernelRequest: hard-linked " << (fromPath ?: "[out-of-root path]") << " -> " << (absolutePath ?: "[out-of-root path]") << " (absolute), ";
-            if (relativeFromPath != nullptr)
-            {
-                cout << "from this root (relative path " << relativeFromPath << ") ";
-            }
-            if (relativePath != nullptr)
-            {
-                cout << "into this root (relative path " << relativePath << ")";
-            }
+            const char* absoluteFromPath = request.paths[MessagePath_From];
+            const char* relativeFromPath = GetRelativePath(absoluteFromPath, s_virtualizationRootFullPath.c_str());
+            
+#ifdef DEBUG
+            cout << "PrjFSLib.HandleKernelRequest: " << (requestHeader->messageType == MessageType_KtoU_NotifyFileHardLinkCreated ? "hard-linked " : "renamed ");
+            cout << "from this root (relative path " << (relativeFromPath == nullptr ? "[NULL]" : relativeFromPath) << ") ";
+            cout << "into this root (relative path " << relativePath << ")";
             cout << endl;
 #endif
             
-            if (relativePath != nullptr)
+            if (strcmp(relativePath, "") == 0)
+            {
+                result = HandleFileNotification(
+                    requestHeader,
+                    relativePath,
+                    absolutePath,
+                    relativeFromPath,
+                    false,
+                    KUMessageTypeToNotificationType(static_cast<MessageType>(requestHeader->messageType)));
+            }
+            else
             {
                 result = HandleNewFileInRootNotification(
                     requestHeader,
                     relativePath,
                     absolutePath,
-                    false, // isDirectory; TODO: may yet be a directory
+                    relativeFromPath,
+                    false,
```

Comment:
Could you add this comment back:

`// isDirectory; TODO: may yet be a directory`

Here and above in the call to `HandleFileNotification`